### ISSUE TITLE
Feature idea: Allow usage on packages that don't have a `py.typed` marker yet

### ISSUE THREAD

**Issue by smheidrich** [Bibo-Joshi/pyright-type-completeness#5]
## Problem

Currently, if the project this Action is run on doesn't have a `py.typed` marker yet, Pyright will [fail](https://github.com/smheidrich/rope/actions/runs/10970870548/job/30465259253) with `error: No py.typed file found`.

But I imagine most library authors that start out with a mostly or completely untyped library will want to only add `py.typed` once it's halfway complete, and in order to get there they might want to use this Action as a "ratcheting" mechanism that ensures monotonically increasing completeness. Which they can't currently do because of the above error.

## Feature idea

In light of the above, it might be nice if this Action provided a way to check the type completeness of projects that don't have a `py.typed` marker yet.

### Option 1: Add `py.typed` marker

The simplest way this might happen is if the Action was able to just add a `py.typed` marker to the project itself if the user configured it to do so. There would probably also have to be a way to specify the package directory under which it should be placed.

### Option 2: Configure Pyright to proceed despite the absence of `py.typed`

I couldn't find any feature of Pyright's that would disable the `py.typed` check, so this would probably involve getting a CLI flag to that effect implemented in Pyright first and then allowing the user of this Action to configure it to be on or off.

Definitely "cleaner" than option (1) and better for the ecosystem as the feature would exist further upstream, but more work.

### Workaround: Users add `py.typed` in their `install-command`s

Users are able to circumvent the need for this feature by setting `install-command` to something like `touch my_package/py.typed && pip install -U .`.

~~But in order for that to actually work, #4 needs to be fixed first, otherwise this will only add `py.typed` to the PR revision but not the base one.~~ ← Never mind that, the reason it didn't work was that I was doing an editable install (`-e`) and had nothing to do with #4.

**Comment by Bibo-Joshi** [comment]
Thanks for laying out the idea 👍Your arguments sound reasonable to me.

I agree that option 2 would be a tad nicer, but I don't see myself opening an issue at pyright for that. If you'd like to do that, I would hold on to this issue here until the pyright request is clarified.

If you do not want to reach out to pyright or if they reject the feature, I'd still happy to accept a PR for option 1. A parameter `py-typed-directory` defaulting to `''` (i.e. don't add) would make sense IMO.

**Comment by smheidrich** [comment]
Thanks for the fast reply! I opened a Pyright issue here: https://github.com/microsoft/pyright/issues/9044

So let's wait and see what they say.

**Comment by smheidrich** [comment]
OK so Pyright won't include this :shrug: 

Regarding the interface this feature could have in pyright-type-completeness, `py-typed-directory` would indeed be the simplest one to implement.

I thought that maybe it would make more sense to just have a boolean `add-py-typed` option, because the package name is already known via the `package-name` option and naively it seems like that should be enough to find the package directory. But that depends on what exactly we mean:

1. If we want to find the directory in the original repo, i.e. prior to installation (e.g. `src/my_package`), that seems to require cooperation from the build backend to tell us where it will get it from. I'm not aware of any standardized way to talk to a build backend and find this information (arguably, it might not even be a meaningful concept, because a build backend may transform the source in arbitrary ways, so that the source might not even have a package dir).
2. If we want to find the location to which the package was installed post-installation, that seems much easier:

    1. Import it and run [`inspect.getfile`](https://docs.python.org/3/library/inspect.html#inspect.getfile) on it. Not ideal because imports may be slow or have side effects (although the latter would be pathological for a library other people are meant to use...).
    2. Use [`importlib.util.find_spec`](https://docs.python.org/3/library/importlib.html#importlib.util.find_spec) (recommended by [an example](https://docs.python.org/3/library/importlib.html#checking-if-a-module-can-be-imported) in the importlib docs), which gives us a [`ModuleSpec`](https://docs.python.org/3/library/importlib.html#importlib.machinery.ModuleSpec), which in turns has attributes like `origin` (points to the package's root `__init__.py` file) and `submodule_search_locations` (will usually have a single element which is the package directory) from which we can determine the package dir.

    So I guess this should work except for edge cases (like namespace packages) and we could then add `py.typed` _after_ the installation to the installed package directory. The only question is whether it's worth the complexity.

I might give both approaches a shot to see how they compare.

**Comment by Bibo-Joshi** [comment]
Thanks for the information and also for listing your alternative options.
Adding py.typed post-init does sound like a convenient way. However, I'm hesitant with the complexity. Using `inspect` and `importlib` in my experience often comes with surprises down the road. For example, I'm wondering how your implementation in #8 would cope with namespace packages. Also, the repository structure of a python package doesn't usually change often. So the automation would only save you from a simple one-time/once-in-a-blue-moon configuration.
Moreover, this action has not yet gained a lot of traction so that I don't like to add too much complexity to it.

In summary, since you've already prepared #9, let's go ahead with that 👍

**Comment by smheidrich** [comment]
I agree that the complexity of #8 ended up being excessive for what it provides. As for #9, I'm having second thoughts as well because the abstraction this provides over just setting an `install-command` that contains `touch my_package/py.typed` is very small. So I would actually put both of these on hold for now and wait to see if any other users would find a feature like this useful.
